include/aws/common/byte_buf.h

```diff
@@ -35,5 +39,147 @@ static inline struct aws_byte_buf aws_byte_buf_from_array(const uint8_t c_str[],
     return (struct aws_byte_buf){.buffer = (uint8_t *)&c_str[0], .len = len};
 }
 
+/**
+ * Represents a movable pointer within a larger binary string or buffer.
+ */
+struct aws_byte_cursor {
+    uint8_t *ptr;
+    size_t len;
+};
+
+static inline struct aws_byte_cursor aws_byte_cursor_from_buf(const struct aws_byte_buf *buf) {
+    return (struct aws_byte_cursor){.ptr = buf->buffer, .len = buf->len};
+}
+
+/**
+ * If index >= bound, bound > (SIZE_MAX / 2), or index > (SIZE_MAX / 2), returns 0.
+ * Otherwise, returns index.  This function is designed to return a value
+ * within [0, bound) even under CPU speculation conditions, and is intended to
+ * be used for SPECTRE mitigation purposes.
+ */
+static inline size_t aws_nospec_index(size_t index, size_t bound) {
+    /*
+     * SPECTRE mitigation - we compute a mask that will be zero if len < 0
+     * or len >= buf->len, and all-ones otherwise, and AND it into the index.
+     * It is critical that we avoid any branches in this logic.
+     */
+
+     /*
+      * We need to know the number of bits in a size_t for this. Unfortunately
+      * there doesn't seem to be a standard define for this...
+      */
+# if (UINTPTR_MAX >> 63 == 1)
+    int size_bits = 64;
+# else
+#  if (UINTPTR_MAX >> 31 == 1)
+    int size_bits = 32;
+#  else
+#   error Unknown number of bits for size_t
+#  endif
+# endif
+
+    /*
+     * Hide the index value from the optimizer. This helps ensure that all this
+     * logic doesn't get eliminated.
+     */
+#if defined(__GNUC__) || defined(__clang__)
+     __asm__ __volatile__("" : "+r" (index));
+#endif
+#if defined(_MSVC_LANG)
+     /*
+      * MSVC doesn't have a good way for us to blind the optimizer, and doesn't
+      * even have inline asm on x64. Some experimentation indicates that this
+      * hack seems to confuse it sufficiently for our needs.
+      */
+     *((volatile uint8_t *)&index) += 0;
+#endif
+
+    /*
+     * If len > (SIZE_MAX / 2), then we can end up with len - buf->len being
+     * positive simply because the sign bit got inverted away. So we also check
+     * that the sign bit isn't set from the start.
+     *
+     * We also check that bound <= (SIZE_MAX / 2) to catch cases where the buffer
+     * is _already_ out of bounds.
+     */
+    size_t negative_mask = index | bound;
+    size_t toobig_mask = bound - index - (uintptr_t)1;
+    size_t combined_mask = negative_mask | toobig_mask;
+
+    /*
+     * combined_mask needs to have its sign bit OFF for us to be in range.
+     * Invert it and sign-extend it across the entire field.
+     *
+     * Note that this technically depends on UB - we need right-shift to
+     * perform sign extension on a negative input - but in practice most
+     * architectures and compilers do this. As an alternative we could do a
+     * multiply, if it were to be necessary.
```

Comment:
My vote is to never depend on UB because:
a) tools like valgrind and clang sanitizer will flag this (hopefully)
b) this lib will be distributed to AWS clients and we shouldn't make assumptions on what machines they will run on.